modules/terraform/hydrate-cluster/monitoring.nix

```diff
@@ -0,0 +1,154 @@
+
+    (check (lib.all allTrue (map (group: if group ? "name" then true else false) nvp.value.groups))
+      ''Declarative alert file has a missing "name" attribute in one of the group list items: ${nvp.name}'')
+
+    (check (lib.all allTrue (map (group: if group ? "rules" then true else false) nvp.value.groups))
+      ''Declarative alert file has a missing "rules" attribute in one of the group list items: ${nvp.name}'')
+
+    (check (lib.all allTrue (map (group: builtins.isList group.rules) nvp.value.groups))
+      ''Declarative alert file contains a group list item with a "rules" attribute that is not a list: ${nvp.name}'')
+
+    (check (lib.all allTrue (map (group:
+      if (builtins.length group.rules) == 0
+      then (builtins.trace ''WARN: Declarative alert file has a group list item with no rules: ${nvp.name}'' true)
+      else true
+    ) nvp.value.groups)) "Undefined error")
+  ];
+
+  # Import the alert files, with interpolation, and check minimal alert syntax requirements
+  importAndLintAlerts = ds: filenames: lib.pipe filenames [
+    (map (file: lib.nameValuePair file (import file (cfg.${ds}.importAttrs or {}))))
+    (map (nvp: lintImport nvp))
+  ];
+
+  filterFiles = suffix: dirs: lib.pipe dirs [
+    (lib.foldl (acc: dir: acc ++ lib.filesystem.listFilesRecursive dir) [])
+    (lib.filter (name: lib.hasSuffix suffix name))
+  ];
+
+  alertKvResource = ds: name: path: let
+    tfName = bittelib.kv.normalizeTfName name;
+  in {
+    name = "vmalert_${ds}_${tfName}";
+    value = {
+      path = "kv/alerts/${ds}/${name}";
+      data_json = var ''file("${path}")'';
+      delete_all_versions = true;
+    };
+  };
+
+  dashKvResource = name: path: let
+    tfName = bittelib.kv.normalizeTfName name;
+  in {
+    name = "grafana_dashboard_${tfName}";
+    value = {
+      path = "kv/dashboards/${name}";
+      data_json = var ''file("${path}")'';
+      delete_all_versions = true;
+    };
+  };
+
+  mkAlertKv = ds: dirs: lib.pipe (filterFiles ".nix" dirs) [
+    checkForNameCollision
+    (importAndLintAlerts ds)
+    (map (nvp: rec { name = filenameToAttrName ".nix" nvp.name; value = builtins.toFile "${name}.json" (builtins.toJSON nvp.value); }))
+    (lib.foldl (acc: v: acc ++ [ (alertKvResource ds v.name v.value) ]) [])
+  ];
+
+  mkDashKv = dirs: lib.pipe (filterFiles ".json" dirs) [
+    checkForNameCollision
+    (map (file: lib.nameValuePair (filenameToAttrName ".json" file) file))
+    (lib.foldl (acc: v: acc ++ [ (dashKvResource v.name v.value) ]) [])
+  ];
+in {
+  tf.hydrate-monitoring.configuration = let
+    tipVmCheck = _: if (cfg ? "vm" && cfg.vm ? "importAttrs") then _ else (builtins.trace tipVmMsg _);
+    tipLokiCheck = _: if (cfg ? "loki" && cfg.loki ? "importAttrs") then _ else (builtins.trace tipLokiMsg _);
+    dashCheck = _: builtins.length config.services.grafana.provision.dashboards == 0;
+    onlyTfDashboards = lib.pipe null [
```

Comment:
The VM module option default of '{}' for importAttrs isn't recognized at the _proto level because this is a submodule option under datasources of modules level declared "vm" and "loki" datasources.  For the importAttrs default to be recognized at the _proto level, the "vm" and "loki" datasources need to be declared there explicitly as well.  As discussed, will add a note to these tips so a user can make them disappear.

Reply:
Removed this block due to the new interface alignment.  Any further added tips will have a statement appended on how to remove them.